keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,705 @@
+- Deprecate the flag
+
+**For non-optional features moving to GA, the graduation criteria must include [conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+-->
+
+TODO
+
+### Upgrade / Downgrade Strategy
+
+Currently, `kubelet --node-ip` has more-or-less no effect when using an
+external cloud provider. (It will require that the IP you pass is a
+known node IP, but it does not do anything else.) Assuming we change it
+to behave the same with external cloud providers as it does with legacy
+cloud providers and bare metal (ie, letting the user choose which node
+IP should be primary), then some users who were passing `--node-ip`
+before would need to stop passing it when upgrading, to avoid getting
+the new behavior. Perhaps this should warn and continue to use the old
+behavior during alpha, then warn and use the new behavior in beta.
+
+(Alternatively, since we need to extend `--node-ip` to support dual
+stack anyway, we could deprecate `--node-ip` and replace it with
+`--node-ips`, where the old `--node-ip` flag would continue to have the
+same semantics on external clouds until it went away.)
+
+Users on bare metal who upgrade and then convert their cluster from
+single-stack to dual-stack might need to switch back to single-stack
+first if they later needed to downgrade, due to the loss of some
+bare-metal-related dual-stack functionality when downgrading.
+
+External programs, scripts, etc that currently look at
+`Node.Status.Addresses` will eventually need to be updated to look at
+the new fields instead.
+
+### Version Skew Strategy
+
+`Node.Status.Addresses` can't actually go away until some hypothetical
+future `corev2.Node`, so older components will still be able to get
+all of the old information from there in a skewed cluster.
+
+During the skew period, components would not be able to assume
+unconditionally that `Node.Status.IPs` would be set, since they might
+be running against an old kubelet or cloud-controller-manager. (This
+includes kubelet itself, in the case of a new kubelet and an old
+cloud-controller-manager.)
+
+## Implementation History
+
+<!--
+Major milestones in the life cycle of a KEP should be tracked in this section.
+Major milestones might include
+- the `Summary` and `Motivation` sections being merged signaling SIG acceptance
+- the `Proposal` section being merged signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+Instead of deprecating `Node.Status.Addresses`, we could just clarify
+its semantics instead. However, this would not fix the "API Cleanliness"
+problems.
```

Comment:
Fwiw I think we can address the confusion issues by just dropping every address other than the "primary internal" address(*).  This is not an API change, since we never guaranteed the presence of all those other addresses, and all the existing code basically just prefers the first InternalIP anyway (because nothing else was ever properly defined).

(*) We need an address for each IP family.  I think an open question is whether we tag a new InternalIPv6 type, or have two InternalIP strings with a value for each family.  I have a mild preference for two InternalIPs because I think it will mean fewer code changes required (only bottom layers, not intermediate layers).

Reply:
> Fwiw I think we can address the confusion issues by just dropping every address other than the "primary internal" address(*). This is not an API change, since we never guaranteed the presence of all those other addresses

I mean, I don't think we ever guaranteed that `Node.Status.Addresses` would have _any_ elements.

There may be users who have workloads that depend on `Node.Status.Addresses` containing multiple addresses. Or there might _not_ be such users. I really don't know. I think our API guarantees force us to err on the side of assuming that if we have been providing this data, then someone might have been consuming it.